src/Controllers/FilesController.js

```diff
@@ -42,22 +44,167 @@ export class FilesController extends AdaptableController {
+      });
+  }
   getMetadata(filename) {
     if (typeof this.adapter.getMetadata === 'function') {
       return this.adapter.getMetadata(filename);
     }
     return Promise.resolve({});
   }
-
+  async updateReferences(data, originalData, className) {
+    const referencesToAdd = [];
+    const referencesToRemove = [];
+    const searchForSubfiles = (keyData, keyOriginalData) => {
+      if (typeof keyData !== 'object') {
+        return;
+      }
+      if (!keyOriginalData) {
+        keyOriginalData = {};
+      }
+      for (const key in keyData) {
+        const val = keyData[key] || {};
+        const original = keyOriginalData[key] || {};
+        if (typeof val !== 'object') {
+          continue;
+        }
+        const { __type, name } = val;
+        if (__type === 'File' || original.__type == 'File') {
+          if (name === original.name) {
+            continue;
+          }
+          if (name && original.name == null) {
+            referencesToAdd.push(name);
+          }
+          if (original.name && name == null) {
+            referencesToRemove.push(name);
+          }
+          continue;
+        }
+        searchForSubfiles(val, original);
+      }
+    };
+    searchForSubfiles(data, originalData);
+    const allFiles = referencesToAdd.concat(referencesToRemove);
+    if (allFiles.length == 0) {
+      return;
+    }
+    const filesToFind = allFiles.map(val => new Parse.File(val).toJSON());
+    const fileQuery = new Parse.Query('_File');
+    fileQuery.containedIn('file', filesToFind);
+    const fileData = await fileQuery.find({ useMasterKey: true });
+    const filesToSave = [];
+    for (const fileObject of fileData) {
+      const { _name } = fileObject.get('file');
+      if (referencesToAdd.includes(_name)) {
+        fileObject.addUnique('references', {
+          objectId: data.objectId,
+          className,
+        });
+      } else {
+        fileObject.remove('references', { objectId: data.objectId, className });
+      }
+      filesToSave.push(fileObject);
+    }
+    await Parse.Object.saveAll(filesToSave, { useMasterKey: true });
+  }
+  async getAuthForFile(config, file, auth, object, className) {
+    if (className === '_File') {
+      return;
+    }
+    const [fileObject] = await config.database.find('_File', {
+      file,
+    });
+    if (!fileObject) {
+      return;
+    }
+    let toSave = false;
+    const tokens = fileObject.tokens || [];
+    const allowed = await this.canViewFile(config, fileObject, auth.user);
+    if (allowed) {
+      const url = file.url.split('?');
+      let appendToken = '';
+      if (url.length > 1) {
+        appendToken = `${url[1]}&`;
+      }
+      const token = randomString(25);
+      appendToken += `token=${token}`;
+      file.url = `${url[0]}?${appendToken}`;
+      const expiry = new Date(new Date().getTime() + 30 * 60000);
+      tokens.push({
+        token,
+        expiry,
+        user: auth.user,
+      });
+      toSave = true;
+    }
```

Comment:
i'm not sure to understand why we have tokens here, are you open to create a quick mind map on how you see the new File system implementation business logic ? 

Free download and free to use, useful for code architecture: https://www.xmind.net/fr/xmind2020/ 😄

Reply:
If we use token may be we should prefer a stateless system like `jsonwebtoken` with a derived hash of the master key as the secret.
(I use this kind of system for the webauthn JWT), it will avoid token management into our code.

```js
const getJwtSecret = config => {
  const hash = crypto.createHash('sha512');
  hash.update(config.masterKey, 'utf-8');
  // Security:
  // sha512 return 128 chars, we can keep only 64 chars since it represent 6,61E98 combinations
  // using the hash allow to reduce risk of compromising the master key
  // if brute force is attempted on the JWT
  return hash.digest().toString('hex').slice(64);
};
```

Reply:
So I am thinking:

If a Parse.Object is fetched with a Parse.File in it, we want to:
1. Check if the `req.user` has correct ACL to view `File`
2. If they do, attach a "token" to the `req.file.url`. Create a DB reference that links `req.user` with the token, with an expiry time.
3. When there is a `GET` request on the file, lookup the Token. If no token exists, throw file not found, as no Parse.Object has been inflated with a Parse.File.
4. Call beforeFind on Parse.File using `token.user` (saved from step 2, so that the file trigger can know the auth)
5. If beforeFind throws, return an error. Otherwise, res.end with the data.

Reply:
> Check if the req.user has correct ACL to view File
> When there is a GET request on the file

Here i see 2 things:
- Security during Parse.Query operation, managed by classic database controller permission check
- Before returning the file url, create a JWT with real file url signed by a derived masterkey, replace the real url with JWT url
- The browser will use the temporary jwt file name ex: https://my-parse.com/file/HereTheJWT
- Parse decode the JWT, get the real file URL, then stream the real file content into response

Here we do not need to re-check the user permission on browser GET request since the JWT could be super short lived (1-2min), and will be always generated ONLY if user has permission to get the object file.

Here a limitation of token system and also jwt file url: It will disable the file browser cache. 
But currently it's not suitable to force User to set up a Parse Session Token cookie.

Note: If a file object is PUBLICLY READABLE, do not generate the JWT and return the url directly. It will allow developers to get file without initiating a query. (ex: avatar account choice at login time).

What do you think about that ? @dblythy

Reply:
The only problem with that is attaching `request.user` in the new `beforeFind(Parse.File)` trigger. I understand what you mean by the JWT token, but if we're going to have a `beforeFind(Parse.File)` trigger run before the GET request, it would be nice to attach which user / masterKey the token is associated with, even for 'public' files (`req.user == null`)

Reply:
I have added a broader explanation in the comments below, let me know any thoughts